sys/src/libc/port/reallocarray.c

```diff
@@ -28,6 +28,17 @@
 void *
 reallocarray(void *optr, size_t nmemb, size_t size)
 {
+	/* if both size or nmemb are 0, and realloc is called with
+	 * 0, it's equivalent to freeing the memory. realloc is even
+	 * allowed to return nil. Hence this function needs to distinguish
+	 * between 'pointer to 0 length array' and 'nil pointer'. In the
+	 * event that nmemb * size is zero, we resize to 1 byte, which
+	 * is usable as a pointer to a 0-length array, and further usable
+	 * in later realloc calls.
+	 */
+	if (nmemb * size == 0)
+		return realloc(optr, 1);
+
 	if ((nmemb >= MUL_NO_OVERFLOW || size >= MUL_NO_OVERFLOW) &&
 	    nmemb > 0 && SIZE_MAX / nmemb < size) {
```

Comment:
Actually, the more I think about it, the more I don't see why that is necessary. The existing code should fall-through to realloc 'realloc()' if either nmemb or size is zero. Looking at our realloc(), it appears that we explicitly catch the 'size == 0' case and return nil. This is technically perfectly legal; according to the C11 standard, section 7.22.3 para 1, "If the size of the space requested is zero, the behavior is implementation-defined: either a null pointer
is returned, or the behavior is as if the size were some nonzero value, except that the
returned pointer shall not be used to access an object." But looking at 'malloc()', it appears that malloc(0) will probably return a real pointer.

Perhaps the fix is to capture the 'size == 0 || nmemb == 0' case in reallocarray() and 'free(optr); return malloc(0);', which should return something non-nil.  E.g.,

        if (size == 0 || nmemb == 0) {
                free(optr);
                return malloc(0);
        }